torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -755,29 +762,33 @@ class AsyncSparseAllreduceWork : public ProcessGroupGloo::AsyncWork {
   std::vector<at::Tensor> allgather_indices(
       const at::Tensor& tensor,
       const std::vector<SparseTensorMetadata>& metadata) {
-    auto max_nnz = metadata[0].nnz();
-    for (size_t i = 1; i < metadata.size(); i++) {
-      max_nnz = std::max(max_nnz, metadata[i].nnz());
+    std::vector<size_t> counts(context->size);
+    size_t totalSize = 0;
+    for (size_t i = 0; i < metadata.size(); i++) {
+      counts[i] = metadata[i].nnz() * tensor.sparse_dim();
+      totalSize += counts[i];
     }
 
-    // There are #sparse_dim() 1-dimensional tensors with nnz elems per rank.
-    auto buffer =
-        at::empty({context->size, tensor.sparse_dim(), max_nnz}, at::kLong);
-    buffer.select(0, context->rank)
-        .narrow(1, 0, tensor._nnz())
-        .copy_(tensor.indices());
+    auto output = at::empty({(int64_t)totalSize}, at::kLong);
 
-    // Allgather indices.
-    gloo::AllgatherOptions opts(context);
-    opts.setOutput(buffer.data<long>(), buffer.numel());
+    // Allgatherv indices.
+    gloo::AllgathervOptions opts(context);
+    opts.setInput(tensor.indices().data<long>(), tensor.indices().numel());
+    opts.setOutput(output.data<long>(), counts);
     opts.setTag(tag);
-    gloo::allgather(opts);
+    gloo::allgatherv(opts);
 
     // Compile indices tensor per rank.
     std::vector<at::Tensor> indices;
     indices.reserve(metadata.size());
+    size_t offset = 0;
     for (size_t i = 0; i < metadata.size(); i++) {
-      indices.push_back(buffer.select(0, i).narrow(1, 0, metadata[i].nnz()));
+      auto nnz = metadata[i].nnz();
+      auto sparseDim = tensor.sparse_dim();
```

Comment:
This can be set as `const` outside the loop (and even at the beginning of the function).